### DISCUSSION TITLE
Module Creator template

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#2332] [General]
The Client project scaffolded by the Module Creator currently creates a folder structure of:

Client
..Modules
....[YourModuleName]
......[razor files]
..Resources
....[YourModuleName]
......[RESX files]

The reason it does this is to provide an indication that a module project could actually contain multiple modules (ie. you could have multiple subfolders beneath the "Modules" folder). However I am curious if this is the right way to communicate this capability, as it makes the structure appear to be quite complicated for new developers. Another approach would be to keep the structure more simple and focused on a single module (and communicate the multiple module capability in other documentation or examples). The simple structure would be:

Client
..[razor files]
..Resources
....[RESX files]

Thoughts?

**Comment by nicpitsch** [comment]
I guess the majority of the solutions will have (sooner or later) more than one module. When propagating the simple directory structure, quite a few developers might be faced with the challenge of having to change it afterwards. Also based on a principle to be consistent I would vote for the current structure to be retained and not to introduce the simple structure.

**Comment by sbwalker** [comment]
One thing I should probably mention is that the hierarchical location of a folder within a project structure does not have any effect on the runtime behavior - this is controller by the namespace attribute in the module:

@namespace [Owner].[Module]

So if a developer started out by creating a module with the simple template and then wanted to add additional modules in the project by creating subfolders and moving razor files, it would not affect the runtime behavior of the original module.

**Comment by larsontim12** [comment]
When we are developing we would use the current structure to keep things organized and consistent. Is it possible that you could choose which one you would want to use?

**Reply by sbwalker** [reply]
The framework supports multiple templates already - however it only ships with one option because of the ongoing maintenance effort. As evidenced by recent PRs, there are challenges in even keeping a single template updated with the latest features. However it would definitely be possible for the community to create and distribute additional templates.

**Comment by markdav-is** [comment]
I think having a multi-module template and a single-module template would be a fine idea.  I've considered building my own flavor of module template with changes to the server structure to simplify, but never took the plunge.  changing the default template you have now sounds confusing, so I'd recommend making multiples.
